src/libos/src/net/socket/unix_socket/stream/socket.rs

```diff
@@ -0,0 +1,190 @@
+        let end = self.end.lock().unwrap();
+        end.as_ref().map(|c| c.set_blocking());
+    }
+
+    // TODO: create the corresponding file in the fs
+    pub fn bind(&self, addr: &UnixAddr) -> Result<()> {
+        if !self.path().is_unnamed() {
+            return_errno!(EINVAL, "the socket is already bound");
+        }
+
+        STREAM_SPACE.add_binder(addr)?;
+
+        *self.path.write().unwrap() = addr.clone();
+        if let Some(ref end) = *self.end.lock().unwrap() {
+            end.set_name(&addr);
+        }
+        Ok(())
+    }
+
+    //TODO: add backlog support
+    pub fn listen(&self, backlog: i32) -> Result<()> {
+        if self.path().is_unnamed() {
+            return_errno!(EINVAL, "the socket is not bound");
+        }
+
+        STREAM_SPACE.add_listener(&self.path())
+    }
+
+    /// A non-blocking accept
+    pub fn accept(&self, flags: FileFlags) -> Result<(Self, Option<UnixAddr>)> {
+        if self.path().is_unnamed() {
+            return_errno!(EINVAL, "the socket is not bound");
+        }
+
+        let sock = STREAM_SPACE.pop_incoming(&self.path())?;
+
+        if flags.contains(FileFlags::SOCK_NONBLOCK) {
+            sock.set_non_blocking();
+        }
+
+        let addr = sock.peer_name().ok();
+
+        debug!("the accepted socket is {:?} from {:?}", sock, addr);
+
+        Ok((sock, addr))
+    }
+
+    /// Backlog is temparily not supported so connect will not block
+    pub fn connect(&self, addr: &Option<UnixAddr>) -> Result<()> {
+        debug!("connect to {:?}", addr);
+
+        if addr.is_none() || addr.contains(&UnixAddr::UnNamed) {
+            return_errno!(EINVAL, "invalid address");
+        }
+
+        if self.is_connected() {
+            return_errno!(EISCONN, "already connected");
+        }
+
+        let path = addr.as_ref().unwrap();
+        if *path == self.path() {
+            return_errno!(EINVAL, "self connect is not supported");
+        }
+
+        let (end_self, end_incoming) = UnixEnd::new_duplex_channel()?;
+        end_incoming.set_name(path);
+        end_self.set_name(&self.path());
+
+        if !self.is_blocking() {
+            end_incoming.set_non_blocking();
+            end_self.set_non_blocking();
+        }
+
+        let incoming_socket = StreamUnix {
+            path: RwLock::new(UnixAddr::UnNamed),
+            end: SgxMutex::new(Some(end_incoming)),
+            is_blocking: AtomicBool::new(self.is_blocking()),
+        };
+
+        STREAM_SPACE.push_incoming(path, incoming_socket)?;
+
+        *self.end.lock().unwrap() = Some(end_self);
+        Ok(())
+    }
+
+    // TODO: handle flags
+    pub fn sendto(&self, buf: &[u8], flags: SendFlags, addr: &Option<UnixAddr>) -> Result<usize> {
+        self.write(buf)
+    }
+
+    // TODO: handle flags
+    pub fn recvfrom(&self, buf: &mut [u8], flags: RecvFlags) -> Result<(usize, Option<UnixAddr>)> {
+        let data_len = self.read(buf)?;
+        let addr = self.peer_name().ok();
```

Comment:
Race condition: There is a chance that `peer_name` may not succeed even the previous `read` succeeds.

Reply:
`ok` converts a `Result` to an `Option`. This will panic. I do not take too much care about this kind of race condition. I do not see those test cases around, either.